File: SortingAlgorithms/MergeSort/merge.py
import random as rand
import os
def rotate(nums,start,end):
    val = nums[end]
    for ind in range(end,start,-1):
        nums[ind] = nums[ind-1]
    nums[start]=val

def merge(nums,left,middle,right):
    #merges 2 sorted ranges in place
    p2 = middle+1
    while left<=middle and p2<=right:
        if nums[left] < nums[p2]:
            left+=1
        else:
            rotate(nums,left,p2)
            left+=1
            p2+=1
            middle+=1

def mergeSort(nums,left=0,right=None):
    if right is None: right = len(nums)-1
    if left<right:
        split = (left+right)//2
        mergeSort(nums,left,split)
        mergeSort(nums,split+1,right)
        merge(nums,left,split,right)

filePath = os.path.dirname(os.getcwd())+"/testinput.txt"
inputData = open(filePath,'r').read()
nums = inputData.strip().split()
nums = [int(num) for num in nums]

mergeSort(nums)

print(nums)

File: SortingAlgorithms/MergeSort/merge.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

//merges two sorted ranges of a vector in place
void merge(vector<int> &nums,int left,int middle,int right){
    int left2 = middle+1;
    while(left<=middle && left2<=right){
        if(nums[left] <= nums[left2]) ++left;
        else{
            rotate(nums.begin()+left,nums.begin()+left2,nums.begin()+left2+1);
            ++left, ++middle, ++left2;
        }
    }
}

void mergeSort(vector<int> &nums, int left, int right){
    if(left<right){
        int splitInd = int((left+right)/2);
        mergeSort(nums,left,splitInd);
        mergeSort(nums,splitInd+1,right);
        merge(nums,left,splitInd,right);
    }
}

int main(){
    ifstream unsortedData ("../testinput.txt");
    vector<int> unsortedInts;
    
    //populate vector with input data
    int temp;
    while(unsortedData>>temp) unsortedInts.push_back(temp);
    unsortedData.close();

    //sort vector
    mergeSort(unsortedInts,0,unsortedInts.size()-1);

    //print vector
    for(int num:unsortedInts)
        cout << num << " ";
    cout<<"\n";
}